### PR TITLE
Add (paths ...) field to (context ...) definition

### PR DESCRIPTION
**Pull request by nojb**
This PR proposes a new field `(add-to-path <string list>)` to the `env` stanza. The strings are interpreted as paths and added to the PATH. This happens before the context is set up, so in particular the resulting PATH is used to look up the compiler tools.

Like `env-vars`, it is only supported on the `workspace` file for now. 

The main motivation is #1692 (allowing to set the compiler path independently of the environment)

### CODE REVIEW

**Comment by diml** [comment]
Allowing to add paths to a build context seems like a good idea. However, if we allow to do this in the `env` stanza, then one is able to modify the path in any dune file. That part feels a bit dodgy to me, or at least it requires to think carefully about how it impacts binaries resolution.

What about adding this field to `Workspace.Context.Common.t` instead?

**Comment by nojb** [comment]
> Allowing to add paths to a build context seems like a good idea. However, if we allow to do this in the `env` stanza, then one is able to modify the path in any dune file. That part feels a bit dodgy to me, or at least it requires to think carefully about how it impacts binaries resolution.

Well, actually this is only interpreted in the `workspace` file (like `env-vars` I think); it is ignored in other `dune` files.

But yes, I agree we can add it to the `context` stanza directly, that's enough for my use case. I will do the change.

**Comment by nojb** [comment]
I did the suggested change and added some words to to the manual about the new field.

**Comment by nojb** [comment]
I wonder if we should support a more general `(add-to-list VAR ...)` to extend any PATH-like variable.

**Comment by nojb** [comment]
(I'm thinking of things like CAML_LD_LIBRARY_PATH)

**Comment by rgrinberg** [comment]
How about having an OSL field for path?

For example: `(paths :standard new-path)`. This seems a bit more consistent with the rest of the dune language.

**Comment by nojb** [comment]
> How about having an OSL field for path?
> 
> For example: `(paths :standard new-path)`. This seems a bit more consistent with the rest of the dune language.

I don't see why not. Here `:standard` is supposed to stand for the existing value of the environment variable, right?

For analogy with `env-vars` and to make it a bit more general, what do you think about :

```
(env-path (PATH :standard new-entry) (CAML_LD_LIBRARY_PATH ...))
```

**Comment by rgrinberg** [comment]
Sounds good to me. Which just need a better name than `env-path`.

**Comment by nojb** [comment]
> Sounds good to me. Which just need a better name than `env-path`.

I went with `(paths ...)` like you suggested.

**Comment by nojb** [comment]
This can be reviewed again.

File: src/context.ml
Comment (diml) [inline src/context.ml]: What about: `Path.of_string s |> Path.to_absolute_filename`? That would make the path be interpreted relative to the workspace root, which is consistent with how other paths are interpreted.

File: src/context.ml
Comment (diml) [inline src/context.ml]: Does this `of_list_reduce` means that we can have the same variable several times inside `(paths ...)`?

i.e.:

```
(paths
 (PATH ...)
 (PATH ...)
)
```
?

That feels dodgy. We should forbid such cases.

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/context.ml
Comment (nojb) [reply src/context.ml]: I thought the current code did  that (isn't `cwd` the workspace root?), but anyway your suggestion is much cleaner.

Comment (diml) [reply src/context.ml]: Yep, the current does that but it seems better to me to leave that job to `Path`

File: src/context.ml
Comment (nojb) [reply src/context.ml]: Yes, it means that; and yes, it is dodgy. Will amend the code to fail in that case.

File: src/context.ml
Comment (nojb) [reply src/context.ml]: Fixed.

File: src/context.ml
Comment (nojb) [reply src/context.ml]: This is now fixed.

**Comment by nojb** [comment]
I think all issues have been addressed.

**Comment by diml** [comment]
Looks good! Could you just move the duplicated variable check to `workspace.ml` directly? This way we get the error right at parsing time. Appart from that it's good to merge.

**Comment by nojb** [comment]
> Looks good! Could you just move the duplicated variable check to `workspace.ml` directly? This way we get the error right at parsing time. Appart from that it's good to merge.

This is done, but the test fails on Windows it seems; I am investigating.

**Comment by nojb** [comment]
Ok, I figured out why it wasn't working on Windows (see last commit) and am happy again. Unless someone objects I will merge once I get the CI to pass.

**Comment by rgrinberg** [comment]
Looks good but please squash the commits or rewrite the history into something that can be followed.

**Comment by nojb** [comment]
> Looks good but please squash the commits or rewrite the history into something that can be followed.

Thanks, I will squash before merging, it is a small patch anyway.

**Comment by nojb** [comment]
Oh no! I meant to click on squash and merge but made a mistake and did a regular merge. Sorry everyone!

**Comment by diml** [comment]
no worries, but yeah, in general let's try to keep master tidy :)

**Pull request merged**
